Fixing the //(x::Number, y::Complex) one liner to accomodate silent overflows and division by zero/infinity

Priynsh opened the pull request:
Fixes #53435 Fixes #56245. I have added conditions for division by zero and infinite handling in the case of complex numbers and overflow handling in the case of integer-based complex numbers. Further, I have added a few tests for the same :)

nhz2 commented on base/rational.jl:
```diff
@@ -98,9 +98,26 @@ function //(x::Rational, y::Rational)
+    return (x//abs2(y))*conj(y)
+end
+function //(x::Number, y::Complex{<:Integer})
+    if isinf(real(y)) || isinf(imag(y))
+        return 0//1
+    end
```
If y is a complex integer isn't `isinf` here always false?

Priynsh replied:
In issue #56245 the errors were thrown for "incorrectly assumes y is finite and nonzero" so I added conditions for for infinite and 0.

kshyatt-aws replied:
Did this get resolved?

Priynsh replied:
Yes! I have changed the implementation now. This was regarding an outdated issue. mb for not resolving.

nhz2 commented on base/rational.jl:
```diff
@@ -98,9 +98,26 @@ function //(x::Rational, y::Rational)
+    end
+    real_y = real(y)
+    imag_y = imag(y)
+    denom = Int32(abs(real_y))^2 + Int32(abs(imag_y))^2
```
Why `Int32` here?

Priynsh replied:
I added that to account for any overflow, but int16 would work perfectly fine as well!

Priynsh replied:
Hey, I’ve implemented scaling to prevent any overflow in calculations (removing any need for INT32. Could you take a look at my approach

nsajko commented:
doctest fails:

```none
│ Subexpression:
│ 
│ 3//2 / (1 + 2im)
│ 
│ Evaluated output:
│ 
│ ERROR: OverflowError: 422212465065984 * 281474976710656 overflowed for type Int64
│ Stacktrace:
│  [1] throw_overflowerr_binaryop(op::Symbol, x::Int64, y::Int64)
│    @ Base.Checked ./checked.jl:163
│  [2] checked_mul
│    @ ./checked.jl:297 [inlined]
│  [3] //
│    @ ./rational.jl:97 [inlined]
│  [4] //(x::Rational{Int64}, y::Complex{Int64})
│    @ Base ./rational.jl:124
│  [5] /(x::Rational{Int64}, y::Complex{Int64})
│    @ Base ./rational.jl:436
│  [6] top-level scope
│    @ none:1
│ 
│ Expected output:
│ 
│ 3//10 - 3//5*im
```

nsajko commented on base/rational.jl:
```diff
@@ -98,9 +98,34 @@ function //(x::Rational, y::Rational)
-
-
+function //(x::Number, y::Complex)
+    if((x//abs2(y))==0//1 || (x//abs2(y))==1//0)
```
Comparing with `iszero`, `isinf`, etc., should be cheaper than with `==`.

Priynsh replied:
implemented!

Priynsh commented:
Hey, the overflow was due to converting complex floats to rationals, I have instead come up with a solution requiring no rational conversions :)

Priynsh commented on base/rational.jl:
```diff
@@ -105,9 +105,33 @@ function //(x::Rational, y::Rational)
+    imag_y = imag(y)
+    m=max(abs(real_y),abs(imag_y))
+    if(m==0)
+        return 1//0
```
these conditions are to make sure tests like 
for y ∈ (1 // 0, -1 // 0)
        @test (7 // complex(y)) == (7 // y)
    end
these pass correctly. 
If m==0 means y = 0 -> 7//0 = 1//0 thus 7//complex(0) should also give 1//0.

Priynsh commented:
okay, so I have added tests from #53453, removed the redundancy related to isinf(y) being true only for rational y, added a divide error for division by complex(0), and removed type instability from function //(x::Number, y::Complex{<:Integer}).

Priynsh commented:
Bump

nhz2 commented on base/rational.jl:
```diff
@@ -105,7 +105,24 @@ function //(x::Rational, y::Rational)
-//(x::Number, y::Complex) = x*conj(y)//abs2(y)
+function //(x::Number, y::Complex)
+    if isinf(y)
+        return complex(x*0)
```
This is type unstable.

You could safely fall back to regular `/` once #53453 is merged.

Priynsh replied:
newer version shouldn't have any instability
